src/engine/BoundManager.h

```diff
@@ -0,0 +1,128 @@
+ ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** BoundManager class is a context-dependent implementation of a centralized
+ ** variable registry and their bounds. The intent it so use a single
+ ** BoundManager object between multiple bound tightener classes, which enables
+ ** those classes to care only about bounds and forget about book-keeping.
+ **
+ ** BoundManager provides a method to obtain a new variable with:
+ ** registerNewVariable().
+ **
+ ** The bound values and tighten flags are stored using context-dependent objects,
+ ** which backtrack automatically with the central _context object.
+ **
+ ** There are two sets of methods to set bounds:
+ **   * set*Bounds     - local method used to update bounds
+ **   * tighten*Bounds - shared method to update bounds, propagates the new bounds
+ **                        to the _tableau (if registered) to keep the assignment and
+ **                      basic/non-basic variables updated accordingly.
+ **
+ ** As soon as bounds become inconsistent, i.e. lowerBound > upperBound, an
+ ** InfeasableQueryException is thrown. In the long run, we want the exception
+ ** replaced by a flag, and switch to the conflict analysis mode instead.
+ **
+ ** It is assumed that variables are not introduced on the fly, and as such
+ ** interaction with context-dependent features is not implemented.
+ **/
+
+#ifndef __BoundManager_h__
+#define __BoundManager_h__
+
+#include "context/cdo.h"
+#include "context/context.h"
+#include "List.h"
+#include "Vector.h"
+
+class Tableau;
+class Tightening;
+class BoundManager
+{
+public:
+    BoundManager( CVC4::context::Context &ctx );
+    ~BoundManager();
+
+    /*
+       Initialize BoundManager and register numberOfVariables of variables
+     */
+    void initialize( unsigned numberOfVariables );
+
+    /*
+       Registers a new variable, grows the BoundManager size and bound vectors,
+       initializes new bounds to +/-inf, and returns the index of the new
+       variable.
+     */
+    unsigned registerNewVariable();
+
+    /*
+       Returns number of registered variables
+     */
+    unsigned getNumberOfVariables();
+
+    /*
+       Communicates bounds to the bound Manager and informs _tableau of the
+       changes, so that any necessary updates can be performed.
+     */
+    bool tightenLowerBound( unsigned variable, double value );
+    bool tightenUpperBound( unsigned variable, double value );
+
+    /*
+       Silently sets bounds to the assigned value, checks bound consistency.
+     */
+    bool setLowerBound( unsigned variable, double value );
+    bool setUpperBound( unsigned variable, double value );
+
+    /*
+       Return current bound value.
+     */
+    double getLowerBound( unsigned variable );
```

Comment:
all getters should be const

Reply:
I am not sure how to achieve this, since the bounds are stored within CDOs that are dynamically allocated, self-managed and need to be handled via non-constant  pointers. All straightforward attempts I made throw cascading errors inside ContextDependentObject class, which would be tricky (if at all possible) to resolve. I will think more about this and talk to the CVC4 people, but this seems likely to be a rabbit-hole.

Reply:
Fixed, Vector::operator[] was the problem the entire time, switched it to get() and it complies.